modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RegisteredDomainProcessor.java

```diff
@@ -0,0 +1,186 @@
+        if (fieldString == null) {
+            return null;
+        }
+        String registeredDomain = SUFFIX_MATCHER.getDomainRoot(fieldString);
+        if (registeredDomain == null) {
+            if (SUFFIX_MATCHER.matches(fieldString)) {
+                return new DomainInfo(fieldString);
+            }
+            return null;
+        }
+        if (registeredDomain.indexOf(".") == -1) {
+            // we have domain with no matching public suffix, but "." in it
+            return null;
+        }
+        return new DomainInfo(registeredDomain, fieldString);
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    private class DomainInfo {
+        private final String registeredDomain;
+        private final String tld;
+        private final String subdomain;
+
+        private DomainInfo(String tld) {
+            this.tld = tld;
+            this.registeredDomain = null;
+            this.subdomain = null;
+        }
+
+        private DomainInfo(String registeredDomain, String domain) {
+            int index = registeredDomain.indexOf(".") + 1;
+            if (index > 0 && index < registeredDomain.length()) {
+                this.tld = registeredDomain.substring(index);
+                this.registeredDomain = registeredDomain;
+                int subdomainIndex = domain.lastIndexOf("." + registeredDomain);
+                if (subdomainIndex > 0) {
+                    this.subdomain = domain.substring(0, subdomainIndex);
+                } else {
+                    this.subdomain = null;
+                }
+            } else {
+                this.tld = null;
+                this.registeredDomain = null;
+                this.subdomain = null;
+            }
+        }
+
+        public String getSubdomain() {
+            return subdomain;
+        }
+
+        public String getRegisteredDomain() {
+            return registeredDomain;
+        }
+
+        public String getTLD() {
+            return tld;
+        }
+    }
+
+    public static final class Factory implements Processor.Factory {
+        @Override
+        public RegisteredDomainProcessor create(
+            Map<String, Processor.Factory> registry,
+            String processorTag,
+            String description,
+            Map<String, Object> config
+        ) throws Exception {
+            String field = ConfigurationUtils.readStringProperty(TYPE, processorTag, config, "field");
+            String targetField = ConfigurationUtils.readStringProperty(TYPE, processorTag, config, "target_field");
+            String targetETLDField = ConfigurationUtils.readOptionalStringProperty(TYPE, processorTag, config, "target_etld_field");
+            String targetSubdomainField = ConfigurationUtils.readOptionalStringProperty(
+                TYPE,
+                processorTag,
+                config,
+                "target_subdomain_field"
+            );
```

Comment:
Shouldn't we default these to ECS field names?

Reply:
The one problem with using ECS field names as defaults in this case is that there isn't one good default to use. These fields that this processor provides are used across at least `dns`, `url`, `destination`, `source`, `server`, and `client` field sets.

Reply:
How about having a single optional `target_field` setting that defines the top-level field under which the ECS fields of `registered_domain`, `subdomain`, and `top_level_domain` would be written. It would default to the blank string which would mean that these fields would be written to the root of the document. If it were set to `dns`, `url`, etc., it could be used to produce ECS-compliant documents for each of those use cases.